ecs-init/engine/engine.go

```diff
@@ -180,15 +189,23 @@ func (e *Engine) StartSupervised() error {
 			return engineError("could not start Agent", err)
 		}
 		log.Infof("Agent exited with code %d", agentExitCode)
-		if agentExitCode == upgradeAgentExitCode {
+
+		switch agentExitCode {
+		case upgradeAgentExitCode:
 			err = e.upgradeAgent()
 			if err != nil {
 				log.Error("could not upgrade agent", err)
+			} else {
+				continue
 			}
+		case terminalFailureAgentExitCode:
+			return errors.New("agent exited with terminal exit code")
+		case terminalSuccessAgentExitCode:
+			return nil
 		}
-	}
-	if agentExitCode == terminalFailureAgentExitCode {
-		return errors.New("agent exited with terminal exit code")
+		d := retryBackoff.Duration()
```

Comment:
If the Agent tickles any of the other cases, that don't cause `ecs-init` to exit, the `retryBackoff` could be excessive and delay the restart of the Agent unnecessarily. 

eg:

1. ECS init takes 3 retries to start up the Agent (`retries = 3`)
2. Agent signals upgrade (`retries = 3`)
3. Agent fails upgrade (`retries = 3`)
4. ECS init delays according to retryBackoff (`retries = 3`)
5. ECS init finally starts the Agent (`retries = 3`)

Can the `retryBackoff` be reset or expire after the Agent has been running for some period of time? Otherwise, customers with transient issues may be left with ECS init configured to *eventually* start the Agent but with a significant and inappropriate delay.

Reply:
I'm not sure I follow. Each time ecs-init wants to start the agent it calls StartSupervised, this creates a new backoff object which will have the number of retries set to 0. Once the agent is running StartSupervised returns nil and the backoff counter is reset.

Reply:
Each time that the loop runs, which in the case of an "upgrade exit" the loop isn't exited, the retry object is reused. So, in the listed scenario, you'd possibly have an Agent take the time at `retries = 3` to start, if that were to be made longer - maybe a longer lived instance or something - then you could have an Agent upgrade fail and take an unnecessarily long time to start up. 

I assume that the `serviceStartMaxRetryTime` is the upper bound and if so, its not a terribly large impact, but it is impact nonetheless.

Reply:
When it exits with "upgrade exit" exit code it will immediately try to upgrade the agent. If the upgrade is successful then the loop continues and doesn't increment the retry counter. The retry counter would only get incremented if the upgrade fails.

Reply:
Yep, that's the scenario I outlined - where the upgrade fails. The point is that its possible to have a retry counter working even after a success case could have reset it.

Reply:
The whole point of this PR is to put a delay in between retries, are you saying that the delay is fundamentally unnecessary in some cases? I guess I'd agree with you there but I'm not clear on how we would determine exactly when it's unnecessary. IMO the delay is not overly onerous as it starts at just 500ms and doubles each retry.

Reply:
Yeah, like I said, right now I think the delay isn't going to become too excessive. As I said above, I think the cases where the retry shouldn't compound are isolated from those where it should. If, for instance, the Agent were to be running for more than `serviceStartMaxRetryTime` then the retry could be reset without introducing a "too fast retry" situation, at least in the purview of `ecs-init`. 

I'm happy to disagree and move forward, these aren't blockers per se, but they should be ack'd and either noted for future improvement or consciously dropped.

Reply:
I'd prefer to leave it as it is. Introducing a running timer introduces a lot of complexity and it also can have unintended effects. If the agent takes some time to return an exit code, during which it was looping over some API calls that we are trying to limit, we might mistakenly assume that the agent has been running for a while and we are safe to run another retry.

That could circumvent what we are trying to achieve, because we don't know how many API calls the agent was running while it was trying to start, nor how long it tried to start for. Keeping the logic as-is enforces a downtime between retries that guarantees that we are not going to spam API calls to CP.

We would essentially just have to hope that the agent returns quickly in order for a running timer to work properly, which I don't think is a safe assumption.